### PR TITLE
Use platform-matrix and split jobs by steps for libraries pipelines

### PR DESCRIPTION
**Pull request by safern**
In this PR we now use platform-matrix and the shared logic to define pool, containers, helix-queues, xplat variables, checkout the repo, etc.

Also, as part of this PR we now split the builds as follows:

1. Build sources
2. Build tests for 3 different OSs (Linux, OSX and Windows).
3. Run tests for our test matrix.

At the beginning this might cause slower builds, but there will be follow up work in order to improve them as we're changing the queues to align with what coreclr uses in order to share the same native toolset and to be coherent with hour dependencies.

Of course we will keep working on improving CI times and making it better. 

This is needed for coreclr to depend on libraries steps for live-live CI and next steps are to start plumbing coreclr into libraries builds.

Also, there's clean-up we can do afterwards to consolidate artifacts names and all the paths we use to publish and download artifacts.

FYI: @stephentoub 

cc: @dotnet/runtime-infrastructure

### CODE REVIEW

**Review by trylek — APPROVED** [review]
Looks awesome to me. Thank you!

**Review by jashook — APPROVED** [review]
This looks great thank you for the work.

File: eng/referenceAssemblies.props
Comment (ViktorHofer) [inline eng/referenceAssemblies.props]: the comment needs to be updated.

File: eng/pipelines/libraries/variables.yml
Comment (ViktorHofer) [inline eng/pipelines/libraries/variables.yml]: isn't this something that coreclr and installer would as well?

Comment (jashook) [reply eng/pipelines/libraries/variables.yml]: It is unclear if it will be used I think

Comment (ViktorHofer) [reply eng/pipelines/libraries/variables.yml]: don't you have any logic specific to internal builds?

Comment (trylek) [reply eng/pipelines/libraries/variables.yml]: Please just be considerate with making additional fundamental changes at this point as it will reset all the testing and delay my live-live pipeline work. Frankly speaking I would be the happiest if we could merge this change in right now its current form and follow up with additional cleanups in separate PR's.

Comment (ViktorHofer) [reply eng/pipelines/libraries/variables.yml]: I'm reviewing the PR right now and will comment on what points out. We can address those in a subsequent PR of course.

Comment (ViktorHofer) [reply eng/pipelines/libraries/variables.yml]: @safern do we in libraries actually still need this variable? Should we instead just use the same pattern that other files use? https://github.com/dotnet/runtime/blob/LibrariesLiveLiveCi/eng/pipelines/common/checkout-job.yml#L11

Comment (jashook) [reply eng/pipelines/libraries/variables.yml]: When going through the review I had similar opinions about isOfficialBuild; however, I think until we stand up the official build we will not know entirely how that will look. So making changes around it now seems like effort incorrectly spent.

Comment (safern) [reply eng/pipelines/libraries/variables.yml]: That was my plan. So at part of this iteration I wanted to just make the minimal set of changes on how we had set it up for libraries (using the same variables and all that). In a follow-up PR I plan working with @trylek and trying to define as much shared variables  as possible. 

But yes, this template can be shared and we can have a variable to know if it is internal or public build. As they use that to define the helix-queues. https://github.com/dotnet/runtime/blob/master/eng/pipelines/coreclr/templates/helix-queues-setup.yml#L118

Comment (safern) [reply eng/pipelines/libraries/variables.yml]: > So making changes around it now seems like effort incorrectly spent.

Yes, that's why I wanted to leave pretty much as is. Don't want to spin up CI again. 

Anyway I will have a follow-up PR probably to unify the test and build steps again for the time being as the machine provisioning delay is impacting the build time by a LOT.

Comment (ViktorHofer) [reply eng/pipelines/libraries/variables.yml]: Thanks.

I believe code which is added via a PR should be reviewed carefully. In my experience, changing or getting rid of stuff after-merge is more costly/complex, mainly because of the lack of context. I would prefer having TODO comments where we intend to change the behavior later but I understand that we want to get this in asap therefore I'm fine with not adding them now.

Comment (ViktorHofer) [reply eng/pipelines/libraries/variables.yml]: This applies to any kind of PR, not specific to this one.

File: eng/pipelines/libraries/base-job.yml
Comment (ViktorHofer) [inline eng/pipelines/libraries/base-job.yml]: why do we need warnaserror false here?

File: eng/pipelines/libraries/variables.yml
Comment (safern) [reply eng/pipelines/libraries/variables.yml]: For this one since the changes were complicated and since we don’t really know what the final shape is going to be, I didn’t even knew which TODOs to add. After this is merged I will go and investigate what can be removed, honestly I submitted it as is as it was blocking @trylek. 

I think reviewing the build result and platform matrix should suffice.

File: eng/pipelines/libraries/base-job.yml
Comment (safern) [reply eng/pipelines/libraries/base-job.yml]: Actually this was ported to xplat-setup and I can remove it. And I also think it was added for a reason I can’t remember anymore. So I will remove it in a follow up PR.

Comment (ViktorHofer) [reply eng/pipelines/libraries/base-job.yml]: Thanks.

> And I also think it was added for a reason I can’t remember anymore.

That's exactly was referring to above with "lack of context" :)

File: eng/pipelines/libraries/build-job.yml
Comment (ViktorHofer) [inline eng/pipelines/libraries/build-job.yml]: I think it's worth discussing if we should extract that into the common layer in a subsequent PR.

Comment (ViktorHofer) [reply eng/pipelines/libraries/build-job.yml]: Related: https://github.com/dotnet/runtime/blob/master/eng/install-nuget-credprovider.sh

Comment (safern) [reply eng/pipelines/libraries/build-job.yml]: Yeah I think this should definitely go into the common layer whenever we do the official build.

File: eng/pipelines/libraries/build-job.yml
Comment (ViktorHofer) [inline eng/pipelines/libraries/build-job.yml]: can we adjust the timeout per leg? Often we needed to change it only for arm/arm64.

**Review by ViktorHofer — APPROVED** [review]

File: eng/pipelines/libraries/build-job.yml
Comment (safern) [reply eng/pipelines/libraries/build-job.yml]: Yes, we can override it in platform-matrix.

**Comment by safern** [comment]
I've modified the PR to run the tests as part of the build step (however the templates for granular steps is already there). The reason why I had to do this is because it turns out our tests are Configuration dependent, so if you build the tests on Debug they test framework behavior for that configuration, which was causing the build to have test failures because we wanted to have the minimal set of test builds. In order to avoid expanding the build-test matrix and introduce more steps, I decided to keep it simple to unblock @trylek's work to be able to merge before the holidays so that he can make progress on his end. 

The plan is to submit a PR to make tests Configuration agnostic and once that's in, I can go back to what I had initially in .azure-ci and I will also do the clean up suggested in the reviews.

**Comment by safern** [comment]
Libraries CI which is the only one really affected is green. Will merge to unblock @trylek's work, I will follow up with PRs once https://github.com/dotnet/runtime/pull/378 is merged.

**Comment by jaredpar** [comment]
>     

Is there an issue tracking the reference assembly code producing `throw null!` instead of disabling this warning?

---
Refers to: eng/referenceAssemblies.props:18 in f985105. [](commit_id = f985105abdd80d7e450e2dc05ad15e5a55fc32fa, deletion_comment = False)

**Comment by jaredpar** [comment]
>     

Or rather if there is no intent to emit `null` correct code in reference assemblies, reasonable, why not use nullable annotate instead of compiling with nullable checking?

---
In reply to: [560437235](https://github.com/dotnet/runtime/pull/274#issuecomment-560437235) [](ancestors = 560437235)

---
Refers to: eng/referenceAssemblies.props:18 in f985105. [](commit_id = f985105abdd80d7e450e2dc05ad15e5a55fc32fa, deletion_comment = False)

**Pull request merged**
